File: trustly-android-sdk/src/test/java/net/trustly/android/sdk/TrustlyJsInterfaceTest.java
package net.trustly.android.sdk;

import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.clearInvocations;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

import net.trustly.android.sdk.views.TrustlyView;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.util.HashMap;

public class TrustlyJsInterfaceTest {

    private static final String EVENT_NAME = "event";

    @Mock
    private TrustlyView mockTrustlyView;

    private TrustlyJsInterface trustlyJsInterface;

    @Before
    public void setUp() {
        MockitoAnnotations.openMocks(this);

        trustlyJsInterface = new TrustlyJsInterface(mockTrustlyView);
    }

    @After
    public void tearDown() {
        clearInvocations(mockTrustlyView);
    }

    @Test
    public void shouldValidateTrustlyJsInterfaceInstanceIsNotNull() {
        assertNotNull(trustlyJsInterface);
    }

    @Test
    public void shouldValidateTrustlyJsInterfacePostMessageWithoutTrustlyViewInstance() {
        trustlyJsInterface = new TrustlyJsInterface(null);
        trustlyJsInterface.postMessage("PayWithMyBank.event|event");
        verify(mockTrustlyView, times(0)).notifyListener(EVENT_NAME, new HashMap<>());
    }

    @Test
    public void shouldValidateTrustlyJsInterfacePostMessageWithoutMessage() {
        trustlyJsInterface.postMessage(null);
        verify(mockTrustlyView, times(0)).notifyListener(EVENT_NAME, new HashMap<>());
    }

    @Test
    public void shouldValidateTrustlyJsInterfacePostMessageNullCommand() {
        trustlyJsInterface.postMessage("|event");
        verify(mockTrustlyView, times(0)).notifyListener(EVENT_NAME, new HashMap<>());
    }

    @Test
    public void shouldValidateTrustlyJsInterfacePostMessageNullCommandAndNullEvent() {
        trustlyJsInterface.postMessage("|");
        verify(mockTrustlyView, times(0)).notifyListener(EVENT_NAME, new HashMap<>());
    }

    @Test
    public void shouldValidateTrustlyJsInterfacePostMessageWithEmptyMessage() {
        trustlyJsInterface.postMessage("");
        verify(mockTrustlyView, times(0)).notifyListener(EVENT_NAME, new HashMap<>());
    }

    @Test
    public void shouldValidateTrustlyJsInterfacePostMessageWithNoValidEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.eventNotValid|event");
        verify(mockTrustlyView, times(0)).notifyListener(EVENT_NAME, new HashMap<>());
    }

    @Test
    public void shouldValidateTrustlyJsInterfacePostMessageWithValidEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|event");
        verify(mockTrustlyView, times(1)).notifyListener(EVENT_NAME, new HashMap<>());
    }

    @Test
    public void shouldValidateTrustlyJsInterfacePostMessageWithNullEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|null");
        verify(mockTrustlyView, times(1)).notifyListener(EVENT_NAME, new HashMap<>());
    }

    @Test
    public void shouldValidateTrustlyJsInterfacePostMessageWithNoPassedEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|");
        verify(mockTrustlyView).notifyListener(EVENT_NAME, new HashMap<>());
    }

    @Test
    public void shouldValidateTrustlyJsInterfacePostMessageWithAllValidEvents() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|event|http://www.url.com|123456|47d7-89d3-9628d4cfb65e|bank_selected|100021|123");
        verify(mockTrustlyView, times(1)).notifyListener(EVENT_NAME, getAllEventNames());
    }

    @Test
    public void shouldValidateTrustlyJsInterfacePostMessageWithoutPageEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|event||123456|47d7-89d3-9628d4cfb65e|bank_selected|100021|123");
        verify(mockTrustlyView, times(1)).notifyListener(EVENT_NAME, getAllOtherEventNames("page"));
    }

    @Test
    public void shouldValidateTrustlyJsInterfacePostMessageWithoutTransactionIdEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|event|http://www.url.com||47d7-89d3-9628d4cfb65e|bank_selected|100021|123");
        verify(mockTrustlyView, times(1)).notifyListener(EVENT_NAME, getAllOtherEventNames("transactionId"));
    }

    @Test
    public void shouldValidateTrustlyJsInterfacePostMessageWithoutMerchantReferenceEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|event|http://www.url.com|123456||bank_selected|100021|123");
        verify(mockTrustlyView, times(1)).notifyListener(EVENT_NAME, getAllOtherEventNames("merchantReference"));
    }

    @Test
    public void shouldValidateTrustlyJsInterfacePostMessageWithoutTypeEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|event|http://www.url.com|123456|47d7-89d3-9628d4cfb65e||100021|123");
        verify(mockTrustlyView, times(1)).notifyListener(EVENT_NAME, getAllOtherEventNames("type"));
    }

    @Test
    public void shouldValidateTrustlyJsInterfacePostMessageWithoutDataEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|event|http://www.url.com|123456|47d7-89d3-9628d4cfb65e|bank_selected||123");
        verify(mockTrustlyView, times(1)).notifyListener(EVENT_NAME, getAllOtherEventNames("data"));
    }

    @Test
    public void shouldValidateTrustlyJsInterfacePostMessageWithoutTransferEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|event|http://www.url.com|123456|47d7-89d3-9628d4cfb65e|bank_selected|100021|");
        verify(mockTrustlyView, times(1)).notifyListener(EVENT_NAME, getAllOtherEventNames("transfer"));
    }

    @Test
    public void shouldValidateTrustlyJsInterfaceResize() {
        trustlyJsInterface.resize(100, 100);
        verify(mockTrustlyView, times(1)).resize(100, 100);
    }

    @Test
    public void shouldValidateTrustlyJsInterfaceAddToListenerDetailsWithAllNullValues() {
        HashMap<String, String> eventDetails = new HashMap<>();
        trustlyJsInterface.addToListenerDetails(null, -1, null, eventDetails);
        assertTrue(eventDetails.isEmpty());
    }

    @Test
    public void shouldValidateTrustlyJsInterfaceAddToListenerDetailsInvalidParamsValue() {
        HashMap<String, String> eventDetails = new HashMap<>();
        trustlyJsInterface.addToListenerDetails(null, 1, EVENT_NAME, eventDetails);
        assertTrue(eventDetails.isEmpty());
    }

    @Test
    public void shouldValidateTrustlyJsInterfaceAddToListenerDetailsInvalidIndexValue() {
        HashMap<String, String> eventDetails = new HashMap<>();
        trustlyJsInterface.addToListenerDetails(new String[] {}, 10, EVENT_NAME, eventDetails);
        assertTrue(eventDetails.isEmpty());
    }

    @Test
    public void shouldValidateTrustlyJsInterfaceAddToListenerDetailsInvalidEventNameValue() {
        HashMap<String, String> eventDetails = new HashMap<>();
        trustlyJsInterface.addToListenerDetails(new String[] {}, 10, null, eventDetails);
        assertTrue(eventDetails.isEmpty());
    }

    @Test
    public void shouldValidateTrustlyJsInterfaceAddToListenerDetailsInvalidEventDetailsValue() {
        HashMap<String, String> eventDetails = null;
        trustlyJsInterface.addToListenerDetails(new String[] {}, 10, null, eventDetails);
        assertNull(eventDetails);
    }

    @Test
    public void shouldValidateTrustlyJsInterfaceAddToListenerDetailsInvalidParamsWithNullValue() {
        HashMap<String, String> eventDetails = new HashMap<>();
        trustlyJsInterface.addToListenerDetails(new String[] { null }, 0, EVENT_NAME, eventDetails);
        assertTrue(eventDetails.isEmpty());
    }

    @Test
    public void shouldValidateTrustlyJsInterfaceAddToListenerDetailsInvalidParamsWithNullStringValue() {
        HashMap<String, String> eventDetails = new HashMap<>();
        trustlyJsInterface.addToListenerDetails(new String[] { "null" }, 0, EVENT_NAME, eventDetails);
        assertTrue(eventDetails.isEmpty());
    }

    private HashMap<String, String> getAllEventNames() {
        HashMap<String, String> eventNames = new HashMap<>();
        eventNames.put("page", "http://www.url.com");
        eventNames.put("transactionId", "123456");
        eventNames.put("merchantReference", "47d7-89d3-9628d4cfb65e");
        eventNames.put("type", "bank_selected");
        eventNames.put("data", "100021");
        eventNames.put("transfer", "123");
        return eventNames;
    }

    private HashMap<String, String> getAllOtherEventNames(String eventName) {
        HashMap<String, String> eventNames = getAllEventNames();
        eventNames.remove(eventName);
        return eventNames;
    }

}

File: trustly-android-sdk/src/test/java/net/trustly/android/sdk/interfaces/TrustlyJsInterfaceTest.kt
package net.trustly.android.sdk.interfaces

import net.trustly.android.sdk.views.TrustlyView
import net.trustly.android.sdk.views.components.TrustlyComponent
import net.trustly.android.sdk.views.events.TrustlyEvents
import org.junit.After
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.mockito.Mock
import org.mockito.Mockito.clearInvocations
import org.mockito.Mockito.times
import org.mockito.Mockito.verify
import org.mockito.MockitoAnnotations

class TrustlyJsInterfaceTest {

    companion object {

        const val EVENT_NAME: String = "event"

    }

    @Mock
    private lateinit var mockTrustlyView: TrustlyView

    @Mock
    private lateinit var mockTrustlyEvents: TrustlyEvents

    private lateinit var trustlyJsInterface: TrustlyJsInterface

    @Before
    fun setUp() {
        MockitoAnnotations.openMocks(this)

        trustlyJsInterface = TrustlyJsInterface(mockTrustlyView, mockTrustlyEvents, TrustlyComponent.Type.WIDGET)
    }

    @After
    fun tearDown() {
        clearInvocations(mockTrustlyView, mockTrustlyEvents)
    }

    @Test
    fun shouldValidateTrustlyJsInterfaceInstanceIsNotNull() {
        assertNotNull(trustlyJsInterface)
    }

    @Test
    fun shouldValidateTrustlyJsInterfacePostMessageWithoutMessage() {
        trustlyJsInterface.postMessage(null)
        verify(mockTrustlyEvents, times(0)).notifyListener(
            EVENT_NAME, HashMap()
        )
    }

    @Test
    fun shouldValidateTrustlyJsInterfacePostMessageNullCommand() {
        trustlyJsInterface.postMessage("|event")
        verify(mockTrustlyEvents, times(0)).notifyListener(
            EVENT_NAME, HashMap()
        )
    }

    @Test
    fun shouldValidateTrustlyJsInterfacePostMessageNullCommandAndNullEvent() {
        trustlyJsInterface.postMessage("|")
        verify(mockTrustlyEvents, times(0)).notifyListener(
            EVENT_NAME, HashMap()
        )
    }

    @Test
    fun shouldValidateTrustlyJsInterfacePostMessageWithEmptyMessage() {
        trustlyJsInterface.postMessage("")
        verify(mockTrustlyEvents, times(0)).notifyListener(
            EVENT_NAME, HashMap()
        )
    }

    @Test
    fun shouldValidateTrustlyJsInterfacePostMessageWithNoDividerMessage() {
        trustlyJsInterface.postMessage("event")
        verify(mockTrustlyEvents, times(0)).notifyListener(
            EVENT_NAME, HashMap()
        )
    }

    @Test
    fun shouldValidateTrustlyJsInterfacePostMessageWithNoValidEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.eventNotValid|event")
        verify(mockTrustlyEvents, times(0)).notifyListener(
            EVENT_NAME, HashMap()
        )
    }

    @Test
    fun shouldValidateTrustlyJsInterfacePostMessageWithValidEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|event")
        verify(mockTrustlyEvents, times(1)).notifyListener(
            EVENT_NAME, HashMap()
        )
    }

    @Test
    fun shouldValidateTrustlyJsInterfacePostMessageWithNullEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|null")
        verify(mockTrustlyEvents, times(1)).notifyListener(
            EVENT_NAME, HashMap()
        )
    }

    @Test
    fun shouldValidateTrustlyJsInterfacePostMessageWithNoPassedEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|")
        verify(mockTrustlyEvents).notifyListener(EVENT_NAME, HashMap())
    }

    @Test
    fun shouldValidateTrustlyJsInterfacePostMessageWithAllValidEvents() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|event|http://www.url.com|123456|47d7-89d3-9628d4cfb65e|bank_selected|100021|123")
        verify(mockTrustlyEvents, times(1)).notifyListener(
            EVENT_NAME, getAllEventNames()
        )
    }

    @Test
    fun shouldValidateTrustlyJsInterfacePostMessageWithoutPageEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|event||123456|47d7-89d3-9628d4cfb65e|bank_selected|100021|123")
        verify(mockTrustlyEvents, times(1)).notifyListener(
            EVENT_NAME, getAllOtherEventNames("page")
        )
    }

    @Test
    fun shouldValidateTrustlyJsInterfacePostMessageWithoutTransactionIdEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|event|http://www.url.com||47d7-89d3-9628d4cfb65e|bank_selected|100021|123")
        verify(mockTrustlyEvents, times(1)).notifyListener(
            EVENT_NAME, getAllOtherEventNames("transactionId")
        )
    }

    @Test
    fun shouldValidateTrustlyJsInterfacePostMessageWithoutMerchantReferenceEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|event|http://www.url.com|123456||bank_selected|100021|123")
        verify(mockTrustlyEvents, times(1)).notifyListener(
            EVENT_NAME, getAllOtherEventNames("merchantReference")
        )
    }

    @Test
    fun shouldValidateTrustlyJsInterfacePostMessageWithoutTypeEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|event|http://www.url.com|123456|47d7-89d3-9628d4cfb65e||100021|123")
        verify(mockTrustlyEvents, times(1)).notifyListener(
            EVENT_NAME, getAllOtherEventNames("type")
        )
    }

    @Test
    fun shouldValidateTrustlyJsInterfacePostMessageWithoutDataEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|event|http://www.url.com|123456|47d7-89d3-9628d4cfb65e|bank_selected||123")
        verify(mockTrustlyEvents, times(1)).notifyListener(
            EVENT_NAME, getAllOtherEventNames("data")
        )
    }

    @Test
    fun shouldValidateTrustlyJsInterfacePostMessageWithoutTransferEvent() {
        trustlyJsInterface.postMessage("PayWithMyBank.event|event|http://www.url.com|123456|47d7-89d3-9628d4cfb65e|bank_selected|100021|")
        verify(mockTrustlyEvents, times(1)).notifyListener(
            EVENT_NAME, getAllOtherEventNames("transfer")
        )
    }

    @Test
    fun shouldValidateTrustlyJsInterfaceResize() {
//        trustlyJsInterface.resize(100f, 100f)
//        verify(mockTrustlyView, times(1)).resize(100f, 100f)
    }

    @Test
    fun shouldValidateTrustlyJsInterfaceAddToListenerDetailsWithAllNullValues() {
        val eventDetails = HashMap<String, String>()
        trustlyJsInterface.addToListenerDetails(null, -1, null, eventDetails)
        assertTrue(eventDetails.isEmpty())
    }

    @Test
    fun shouldValidateTrustlyJsInterfaceAddToListenerDetailsInvalidParamsValue() {
        val eventDetails = HashMap<String, String>()
        trustlyJsInterface.addToListenerDetails(null, 1, EVENT_NAME, eventDetails)
        assertTrue(eventDetails.isEmpty())
    }

    @Test
    fun shouldValidateTrustlyJsInterfaceAddToListenerDetailsInvalidIndexValue() {
        val eventDetails = HashMap<String, String>()
        trustlyJsInterface.addToListenerDetails(arrayOf(), 10, EVENT_NAME, eventDetails)
        assertTrue(eventDetails.isEmpty())
    }

    @Test
    fun shouldValidateTrustlyJsInterfaceAddToListenerDetailsInvalidEventNameValue() {
        val eventDetails = HashMap<String, String>()
        trustlyJsInterface.addToListenerDetails(arrayOf(), 10, null, eventDetails)
        assertTrue(eventDetails.isEmpty())
    }

    @Test
    fun shouldValidateTrustlyJsInterfaceAddToListenerDetailsInvalidEventDetailsValue() {
        val eventDetails: HashMap<String, String>? = null
        trustlyJsInterface.addToListenerDetails(arrayOf(), 10, null, eventDetails)
        assertNull(eventDetails)
    }

    @Test
    fun shouldValidateTrustlyJsInterfaceAddToListenerDetailsInvalidParamsWithNullStringValue() {
        val eventDetails = HashMap<String, String>()
        trustlyJsInterface.addToListenerDetails(arrayOf("null"), 0, EVENT_NAME, eventDetails)
        assertTrue(eventDetails.isEmpty())
    }

    private fun getAllEventNames() = hashMapOf(
        "page" to "http://www.url.com",
        "transactionId" to "123456",
        "merchantReference" to "47d7-89d3-9628d4cfb65e",
        "type" to "bank_selected",
        "data" to "100021",
        "transfer" to "123"
    )

    private fun getAllOtherEventNames(eventName: String): HashMap<String, String> {
        val eventNames = getAllEventNames()
        eventNames.remove(eventName)
        return eventNames
    }

}